File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if digits == "":
            return []

        letters = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }
        answer = []
        def dfs(target, i, comb):
            if len(target) == i:
                answer.append(comb)
                return
            

            for char in letters[digits[i]]:
                dfs(target, i+1, comb + char)

        dfs(digits, 0, "")

        return answer

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
    private List<String> answer = new ArrayList<>();
    private Map<String, String> hashMap = new HashMap<>();

    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0){
            return answer;
        }
        hashMap.put("2", "abc");
        hashMap.put("3", "def");
        hashMap.put("4", "ghi");
        hashMap.put("5", "jkl");
        hashMap.put("6", "mno");
        hashMap.put("7", "pqrs");
        hashMap.put("8", "tuv");
        hashMap.put("9", "wxyz");

        dfs(digits, 0, "");
        return answer;

    }

    void dfs(String num, int index, String comb){
        if (index == num.length()){
            answer.add(comb);
            return;
        }
        String str = hashMap.get(String.valueOf(num.charAt(index)));
        for (int i = 0; i < str.length(); i++){
            dfs(num, index+1, comb + String.valueOf(str.charAt(i)));
        }
    }
}